### PR TITLE
Expose ngram_indicies and subword_indices, add test module

### PR DESCRIPTION
**Pull request by RealNicolasBourbaki**

### CODE REVIEW

File: src/vocab.rs [hunk 1]
```diff
@@ -34,6 +35,39 @@ impl PyVocab {
...
+                .unwrap()
+                .into_iter()
+                .map(|pair| pair.into_py(gil.python()))
+                .collect()),
+            VocabWrap::FinalfusionSubwordVocab(inner) => Ok(inner
+                .ngram_indices(word)
+                .unwrap()
+                .into_iter()
+                .map(|pair| pair.into_py(gil.python()))
+                .collect()),
```
Comment (RealNicolasBourbaki) [inline src/vocab.rs]: I don't think this part looks good because of the repeat of code, so I also tried using `macro` to create syntax that allows matching `FinalfusionSubwordVocab(inner)` and `FastTextSubwordVocab(inner)` compactly. But I'm not so sure which way is better.

File: src/vocab.rs [hunk 2]
```diff
@@ -34,6 +35,39 @@ impl PyVocab {
             }
         })
     }
+
+    fn ngram_indices(&self, word: &str) -> PyResult<Vec<Py<PyTuple>>> {
```
Comment (sebpuetz) [inline src/vocab.rs]: `(String, usize)` implements direct conversion into `PyTuple`, so the return type should be `PyResult<Option<Vec<(String, usize)>>>`. 
I would also prefer returning an `Option<Vec<_>>` because explicit NGram vocabs could be unable to produce `(ngram, index)` tuples.

File: src/vocab.rs [hunk 3]
```diff
@@ -34,6 +35,39 @@ impl PyVocab {
...
+        let embeds = self.embeddings.borrow();
+        let gil = pyo3::Python::acquire_gil();
+        match embeds.vocab() {
+            VocabWrap::FastTextSubwordVocab(inner) => Ok(inner
+                .ngram_indices(word)
+                .unwrap()
+                .into_iter()
+                .map(|pair| pair.into_py(gil.python()))
+                .collect()),
```
Comment (sebpuetz) [inline src/vocab.rs]: These lines won't be needed if the method header is changed:
```Rust
                .unwrap()
                .into_iter()
                .map(|pair| pair.into_py(gil.python()))
                .collect()),
```

File: src/vocab.rs (hunk 1 above)
Comment (sebpuetz) [reply src/vocab.rs]: After calling `ngram_indices` on both vocab types, you end up with the same type, `Option<Vec<(String, usize)>>`, so starting from that point you could return the result from those calls from the match and assign it to a variable.

If we stick to my suggestions from above, then the code-duplication would be eliminated anyways, since both match arms would consist of `Ok(inner.ngram_indices(word))`, I don't believe there is a more concise way to write that.

File: src/vocab.rs [hunk 4]
```diff
@@ -34,6 +35,39 @@ impl PyVocab {
...
+            VocabWrap::FinalfusionSubwordVocab(inner) => Ok(inner
+                .ngram_indices(word)
+                .unwrap()
+                .into_iter()
+                .map(|pair| pair.into_py(gil.python()))
+                .collect()),
+            _ => Err(exceptions::ValueError::py_err(
```
Comment (sebpuetz) [inline src/vocab.rs]: Catch-all match arms can be quite hard to debug since the compiler doesn't complain when a new variant is added to the enum. We might end up adding another `VocabType` and we wouldn't be aware of all the places in the code where we need to handle things differently.
 So, when matching enums or something that could change from our side, we should always use explicit match arms. In this case, `VocabWrap::SimpleVocab(_)`.

File: src/vocab.rs [hunk 5]
```diff
@@ -34,6 +35,39 @@ impl PyVocab {
...
+                "subwords-indices queries are not supported for vocabularies of type 'SimpleVocab'",
+            )),
+        }
+    }
+
+    fn subword_indices(&self, word: &str) -> PyResult<Vec<usize>> {
```
Comment (sebpuetz) [inline src/vocab.rs]: Similar to the method above, we should return `PyResult<Option<_>>` here.

File: src/vocab.rs [hunk 6]
```diff
@@ -34,6 +35,39 @@ impl PyVocab {
...
+        let embeds = self.embeddings.borrow();
+        match embeds.vocab() {
+            VocabWrap::FastTextSubwordVocab(inner) => Ok(inner.subword_indices(word).unwrap()),
+            VocabWrap::FinalfusionSubwordVocab(inner) => Ok(inner.subword_indices(word).unwrap()),
+            _ => Err(exceptions::ValueError::py_err(
```
Comment (sebpuetz) [inline src/vocab.rs]: Same issue with `_ => {...}`, explicit matching would be nicer.

**Review by sebpuetz — CHANGES_REQUESTED** [review]
Looks good, but there are some comments!

**Review by sebpuetz — APPROVED** [review]
Looks good to me, but let's wait for @danieldk's opinion

File: src/vocab.rs [hunk 7]
```diff
@@ -34,6 +34,28 @@ impl PyVocab {
...
         })
     }
+
+    fn ngram_indices(&self, word: &str) -> PyResult<Option<Vec<(String, usize)>>> {
+        let embeds = self.embeddings.borrow();
+        match embeds.vocab() {
+            VocabWrap::FastTextSubwordVocab(inner) => Ok(inner.ngram_indices(word)),
+            VocabWrap::FinalfusionSubwordVocab(inner) => Ok(inner.ngram_indices(word)),
+            VocabWrap::SimpleVocab(_) => Err(exceptions::ValueError::py_err(
+                "ngram-indices queries are not supported for vocabularies of type 'SimpleVocab'",
```
Comment (danieldk) [inline src/vocab.rs]: "ngram-indices queries are not supported for vocabularies of type 'SimpleVocab'" -> "querying n-gram indices is not supported for this vocabulary"

The Rust type is an implementation detail.

It is currently not possible for the caller to know this. So maybe it would be nice to add a method `is_subword_vocab`. Though this can go to a separate PR.

File: src/vocab.rs [hunk 8]
```diff
@@ -34,6 +34,28 @@ impl PyVocab {
...
+        }
+    }
+
+    fn subword_indices(&self, word: &str) -> PyResult<Option<Vec<usize>>> {
+        let embeds = self.embeddings.borrow();
+        match embeds.vocab() {
+            VocabWrap::FastTextSubwordVocab(inner) => Ok(inner.subword_indices(word)),
+            VocabWrap::FinalfusionSubwordVocab(inner) => Ok(inner.subword_indices(word)),
+            VocabWrap::SimpleVocab(_) => Err(exceptions::ValueError::py_err(
+                "subwords' indices queries are not supported for vocabularies of type 'SimpleVocab'",
```
Comment (danieldk) [inline src/vocab.rs]: Same here...

File: Cargo.toml [hunk 9]
```diff
@@ -21,7 +21,7 @@ features = ["extension-module"]
 [dependencies]
 itertools = "0.8"
 failure = "0.1"
-finalfusion = "0.8"
+finalfusion = {git = "https://github.com/finalfusion/finalfusion-rust.git", rev = "da4e8d6"}
```
Comment (danieldk) [inline Cargo.toml]: We should postpone merging of this PR until we have a stable `finalfusion` version that has the necessary API changes. We should avoid using git revisions too often.

I will check if we can merge the necessary changes without API breakage (we could then do a minor release).

**Review by danieldk — CHANGES_REQUESTED** [review]
Some more change requests:

* Do not add a 80MB embedding file to the repository! Data files should be 50-100kb max. There are two possibilities here: (1) check if the finalfusion API provides enough to construct a small bogus embedding matrix; this would be ideal, because it does not require any additional data; (2) train embeddings with finalfrontier with a high enough mincount and low enough bucket exponent to get in the 50kb-100kb range.

* Commit messages should be clean. The commit message contains a typo (ngram_indicies). Adding tests should be natural. So the commit message should e.g. be: "Add ngram_indices and subword_indices" or "Add methods to compute subword ngrams and indices" and describe the method names in the commit message body.

File: Cargo.toml (hunk 9 above)
Comment (danieldk) [reply Cargo.toml]: Uploaded `finalfusion` 0.8.2 to crates.io, which adds the necessary traits. So, if you bump the version here to 0.8.2, it should work.

**Review by danieldk — APPROVED** [review]
